core/inc/knoting/networking_types.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+#include <knoting/types.h>
+#include <yojimbo.h>
+
+namespace knot {
+using namespace yojimbo;
+
+constexpr int SERVER_PORT = 13189;
+constexpr int CLIENT_PORT = 12646;
+constexpr int MAX_CLIENTS = 6;
+const char* const SERVER_ADDRESS = "100.117.241.88";
+// const char* const SERVER_ADDRESS = "127.0.0.1";
+//^^^ FOR LOCAL HOST TESTING ^^^
+const double TICK = 1.0 / 30.0;
+
+struct GameConfig : public ClientServerConfig {
+    GameConfig() {
+        ClientServerConfig();
+        this->numChannels = 2;
+        this->channel[0].type = yojimbo::CHANNEL_TYPE_RELIABLE_ORDERED;
+        this->channel[1].type = yojimbo::CHANNEL_TYPE_UNRELIABLE_UNORDERED;
+        this->timeout = 5;
+        this->networkSimulator = false;
+        this->maxPacketSize = 1500;
+    }
+};
+
+template <typename Stream>
+bool serialize_vec2i(Stream& stream, vec2i& val) {
+    auto valX = (int16_t)val.x;
+    auto valY = (int16_t)val.y;
+    serialize_int(stream, valX, -32767, 32767);
+    serialize_int(stream, valY, -32767, 32767);
+    val.x = valX;
+    val.y = valY;
+    return true;
+}
+
+class ClientMessage : public Message {
+   public:
+    ClientMessage() : m_sequence(0), m_recentAck(0) {
+        m_lookAxis = vec2i();
+        m_moveAxis = vec2i();
+        jumpPressed = false;
+        isShooting = false;
+    }
+
+    template <typename Stream>
+    bool Serialize(Stream& stream) {
+        serialize_bits(stream, m_sequence, 16);
+        serialize_bits(stream, m_recentAck, 16);
+
+        serialize_vec2i(stream, m_lookAxis);
+        serialize_vec2i(stream, m_moveAxis);
+        serialize_bool(stream, jumpPressed);
+        serialize_bool(stream, isShooting);
+        return true;
+    }
+
+    vec2i m_lookAxis;
+    vec2i m_moveAxis;
+
+    bool jumpPressed;
+    bool isShooting;
+
+    uint16_t get_sequence() { return m_sequence; }
+    void set_sequence(uint16_t seq) { m_sequence = seq; }
+
+    uint16_t get_recent_ack() { return m_recentAck; }
+    void set_ack(uint16_t reAck) { m_recentAck = reAck; }
+    YOJIMBO_VIRTUAL_SERIALIZE_FUNCTIONS();
+
+   protected:
+    uint16_t m_sequence;
+    uint16_t m_recentAck;
+};
+
+template <typename Stream>
+bool serialize_vec3(Stream& stream, vec3& val) {
+    serialize_float(stream, val.x);
+    serialize_float(stream, val.y);
+    serialize_float(stream, val.z);
+    return true;
+}
+
+template <typename Stream>
+bool serialize_quat(Stream& stream, quat& val) {
+    serialize_float(stream, val.x);
+    serialize_float(stream, val.y);
+    serialize_float(stream, val.z);
+    serialize_float(stream, val.w);
+    return true;
+}
+
+constexpr uint16_t MAX_COLLISIONS = 20;
```

Comment:
is 20 collisions enough between 6 players or is this 20 per client?

Reply:
The vector associated with this will pass back the positions of up to the last 20 collisions of a paint bullet with the environment for client side drawing of paint on surfaces. This is it's only purpose as all other collision handling will happen server side.
20 should be enough, but can't test until more code has been written